### PR TITLE
SoA-Data-Layout-Abstraction

### PR DESCRIPTION
**Pull request by pranavich18**
## Description
GSoC'22 Project
Attempts to create a Structure of Arrays (SoA) Data layout for MSSpectrum class using Zero-Cost Abstraction.

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

File: src/openms/include/OpenMS/KERNEL/MSSpectrum.h [hunk 1]
```diff
@@ -605,6 +610,10 @@ namespace OpenMS
     PeakType::IntensityType calculateTIC() const;
 
 protected:
+
+    /// Peak1d Tuple
+    Container spectra();
```
Comment (jpfeuffer) [inline src/openms/include/OpenMS/KERNEL/MSSpectrum.h:615]: it's peaks. You are in the spectrum class. A spectrum does not contain multiple spectra

File: src/openms/include/OpenMS/KERNEL/Peak1DT.h [hunk 2]
```diff
@@ -0,0 +1,396 @@
+// SOA and AOS tests on god bolt
```
Comment (jpfeuffer) [inline src/openms/include/OpenMS/KERNEL/Peak1DT.h:1]: I would probably call that file differently. Something like MSSpectrumAoSWrapper.h

File: src/openms/include/OpenMS/KERNEL/MSSpectrum.h [hunk 3]
```diff
@@ -104,6 +105,10 @@ namespace OpenMS
     //@{
     /// Peak type
     typedef OpenMS::Peak1D PeakType;
+    /// Peak1D Tuple
+    typedef Peak1DT<double, double> Peak1DTuple;
+    /// Container for Peak1DT
+    typedef BaseContainer<MyVector, Peak1DTuple> Container;
```
Comment (jpfeuffer) [inline src/openms/include/OpenMS/KERNEL/MSSpectrum.h:111]: Find better name for MyVector. E.g. VectorTemplate

**Review by jpfeuffer — COMMENTED** [review]
https://github.com/OpenMS/OpenMS/pull/6267/files#diff-e8dde25859610a8165712f483bebf687b84f5a1b5b2f6089cf5711875592ec12L67
This has to go. A Spectrum should not be a vector of Peak structs anymore

File: src/openms/include/OpenMS/KERNEL/MSSpectrum.h (hunk 1 above)
Comment (pranavich18) [reply src/openms/include/OpenMS/KERNEL/MSSpectrum.h:615]: Sure I'll correct that.

**Comment by pranavich18** [comment]
> https://github.com/OpenMS/OpenMS/pull/6267/files#diff-e8dde25859610a8165712f483bebf687b84f5a1b5b2f6089cf5711875592ec12L67 This has to go. A Spectrum should not be a vector of Peak structs anymore

Are you referring to `#include<Peak1D.h>` ??

Actually I didn't remove in the very beginning to avoid any code breaks.

**Comment by pranavich18** [comment]
> Yes, plus the fact that you cannot inherit from std::vector anymore. I don't think you have to care about code breaks. It will break anyway for some time.

Sure, I'll correct that.

File: src/openms/include/OpenMS/KERNEL/MSSpectrumSoAWrapper.h [hunk 4]
```diff
@@ -137,12 +189,58 @@ struct BaseContainer
         resize(size_);
     }
 
+    bool empty()
+    {
+        if(this->size()==0)
+            return true;
+        
+        return false;
```
Comment (jpfeuffer) [inline src/openms/include/OpenMS/KERNEL/MSSpectrumSoAWrapper.h:197]:
```suggestion
        return this->size() == 0;
```

or even `!this->size()`

File: src/openms/include/OpenMS/KERNEL/MSSpectrumSoAWrapper.h [hunk 5]
```diff
@@ -137,12 +189,58 @@ struct BaseContainer
...
+    void reserve( size_t size_ )
+    {
+        policy_t::reserve(mValues, size_);
+    }
+
+    iterator erase( iterator it )
+    {
+        std::size_t position_ = it.getPosition();
+        std::size_t newPos_ = policy_t::erase( mValues, position_ );
+        return (this->begin() + newPos_);
```
Comment (jpfeuffer) [inline src/openms/include/OpenMS/KERNEL/MSSpectrumSoAWrapper.h:226]: check that newPos is not bigger or equal to size. if it is, return this->end()

**Pull request closed without merging**